web_src/js/components/ActivityHeatmap.vue

```diff
@@ -74,6 +64,16 @@ export default {
 
       const newSearch = params.toString();
       window.location.search = newSearch.length ? `?${newSearch}` : '';
+    },
+    tooltipFormatter(v, locale) {
+      // TODO: use the localized number as below (why is it throwing Uncaught DOMException: Failed to execute
+      // 'attachShadow' on 'Element': Shadow root cannot be created on a host which already hosts a shadow tree.?)
+      // const number = `<gitea-locale-number data-number="${v.count}">${v.count}</gitea-locale-number>`;
```

Comment:
@wxiaoguang I really wanted to use #23861 but I think there's a bug with the detaching logic

Reply:
I think it's time to do more work for these TODOs:

Although I was thinking about leaving it to the complete ICU message refactoring time, it's not bad to take this small step now.

I will try.

![image](https://user-images.githubusercontent.com/2114189/232177867-73da8c04-cf87-4758-a718-daabac01227d.png)

Reply:
See:  Refacotr locale number #24134

Reply:
So it'll be entirely in the backend, I need to render on the frontend, I'll just do it with `formatNumber`

Reply:
You can format the message from backend, then just display it in frontend, eg:

```
ctx.Data["HeatmapTotalContributionsText"] = locale.TrN(  ) // sooner or later , we need to handle the plural correctly
```

```
<div data-text-total-contributions={{.HeatmapTotalContributionsText}}>
```

Reply:
This is pretty much done, look at the code of this conversation, I'm talking about the tooltip

Reply:
Oh I see, for `tooltipFormatter`,  TBH, `<b>%[1]s contributions</b> on %[2]s` is not right in most languages.

Some languages only have one unique form "contribution", some have others like "contribution-for-0", "contribution-for-2", etc.

So I do not think it could be done like this in frontend.

At least, we need to introduce ICU message format first, and let frontend use a ICU messgea formatter to get the correct texts.

Reply:
Let's land like this and refactor when we have the ICU message format

Reply:
But it makes difficulty for translators .... they would complain about such untranslatable strings.

Reply:
We have many strings like this...
![image](https://user-images.githubusercontent.com/20454870/232196138-7f599640-9d69-49f3-a238-4dec85379995.png)

Reply:
Yup, at least, `TrN` is much better than the only-one `number_of_contributions_on_date`, there is a `trNLangRules` , not perfect but it works for most cases.

Reply:
I'll try to replicate on the frontend

Reply:
https://github.com/go-gitea/gitea/pull/24131/commits/90fc40f4fdbb99bacd9e7c0d4f0485a04984196a

Reply:
I do not think it works or is maintainable ..... 

The `Tr` (`fmt.Sprintf`) expects the arguments to match the `%d` parameters. Otherwise the output would be: `the value is %d` => `the value is %!d(MISSING)`

Reply:
I'm not doing it in the backend

Reply:
You are calling `{{.locale.Tr "heatmap.number_of_contributions_on_date_n"}}`

Do the parameters & arguments match?

Reply:
I'm not trying to do any replacements in the backend. It works. See screenshots in PR description

Reply:
Hmm .... I see there is a trick on backend, the `Tr` ignores the empty arguments list, it's not 100% the same as Sprintf  ....  tricky, so finally you get the original strings.

Reply:
![image](https://user-images.githubusercontent.com/20454870/232199255-41e133c6-8c4b-4e5f-99ef-023c916db02e.png)